pkg/generators/openapi.go

```diff
@@ -662,3 +674,95 @@ func (g openAPITypeWriter) generateSliceProperty(t *types.Type) error {
 	g.Do("},\n},\n},\n", nil)
 	return nil
 }
+
+// CustomArgs is used by the gengo framework to pass args specific to the API linter
+type CustomArgs struct {
+	ReportFilename string
+}
+
+type apiLinter struct {
+	rules          []APIRule
+	violations     []apiViolation
+	reportFilename string
+}
+
+func newAPILinter(reportFilename string) *apiLinter {
+	return &apiLinter{
+		rules: []APIRule{
+			&rules.NamesMatch{},
+		},
+		reportFilename: reportFilename,
+	}
+}
+
+type apiViolation struct {
+	// Name of rule from APIRule.Name()
+	rule string
+
+	packageName string
+	typeName    string
+
+	// Optional: name of field that violates API rule. Empty fieldName implies that
+	// the entire type violates the rule.
+	field string
+}
+
+// APIRule is the interface for validating API rule on Go types
+type APIRule interface {
+	// Validate evaluates API rule on type t and returns a list of field names in
+	// the type that violate the rule. Empty field name [""] implies the entire
+	// type violates the rule.
+	Validate(t *types.Type) ([]string, error)
+
+	// Name returns the name of APIRule. The name should not contain ','.
+	Name() string
+}
+
+func (l *apiLinter) validate(t *types.Type) error {
+	for _, r := range l.rules {
+		glog.V(5).Infof("validating API rule %v for type %v", r.Name(), t)
+		fields, err := r.Validate(t)
+		if err != nil {
+			return err
+		}
+		for _, field := range fields {
+			l.violations = append(l.violations, apiViolation{
+				rule:        r.Name(),
+				packageName: t.Name.Package,
+				typeName:    t.Name.Name,
+				field:       field,
+			})
+		}
+	}
+	return nil
+}
+
+func (g *openAPIGen) Finalize(c *generator.Context, w io.Writer) error {
+	// If report file isn't specified, log and return error
+	if len(g.linter.reportFilename) == 0 {
+		for _, v := range g.linter.violations {
+			glog.Errorf("API rule violation: %s,%s,%s,%s", v.rule, v.packageName, v.typeName, v.field)
+		}
+		if len(g.linter.violations) > 0 {
+			return fmt.Errorf("API rule violations exist")
+		}
+		return nil
+	}
+	// Otherwise, print the error to report file and return nil
+	f, err := os.Create(g.linter.reportFilename)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	for _, v := range g.linter.violations {
+		fmt.Fprintf(f, "API rule violation: %s,%s,%s,%s\n", v.rule, v.packageName, v.typeName, v.field)
+	}
+	if len(g.linter.violations) > 0 {
+		// NOTE: we don't return error here because we assume that the report file will
+		// get evaluated afterwards to determine if error should be raised. For example,
+		// you can have make rules that compare the report file with existing known
+		// violations (whitelist) and determine no error if no change is detected.
+		glog.Errorf("API rule violations exist; see %s for details.", g.linter.reportFilename)
```

Comment:
I'm puzzled about the need for this log message, especially if the errors went to stdout.

Reply:
Please see comments below for: 
> what happens if the reportFilename == ""?

Reply:
why not set a list of errors on the openAPIGen struct instead of writing them to a file here?

Reply:
Also, why not log to stderr?

Reply:
> why not set a list of errors on the openAPIGen struct instead of writing them to a file here?

In kubernetes use case we want to achieve two things: 1. the openapi spec generation process should not stop even if violations exist; 2. some Make rule / script should be able to compare the new detected violations and checked-in known violations and make sure that the list don't get bigger.

That essentially means we cannot return error in any of the `openAPIGen.Init()`, `openAPIGen.GenerateType()`, `openAPIGen.Finalize()` without changing `gengo` framework, because [`assembler.AssembleFile()` won't run](https://github.com/kubernetes/gengo/blob/9667b67e4b895d34f09fad596d28f285e0197131/generator/execute.go#L275-L291) to generate openapi spec file.

> Also, why not log to stderr?

Removed the `glog` line. One question is that we currently use `--report-filename=-` to indicate printing violations to `stdout`. Do we have some UNIX convention other than `"-"` for `stderr` if we want to print violations to `stderr`?